src/sbvr-api/permissions.ts

```diff
@@ -0,0 +1,1058 @@
+															}
+														}]
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	return getPermissions(permsFilter).nodeify(callback)
+}
+
+const $getApiKeyPermissions = memoize(
+	(apiKey: string) => {
+		const permsFilter = {
+			$or: {
+				is_of__api_key: {
+					$any: {
+						$alias: 'khp',
+						$expr: {
+							khp: {
+								api_key: {
+									$any: {
+										$alias: 'k',
+										$expr: {
+											k: { key: apiKey }
+										}
+									}
+								}
+							}
+						}
+					}
+				},
+				is_of__role: {
+					$any: {
+						$alias: 'rhp',
+						$expr: {
+							rhp: {
+								role: {
+									$any: {
+										$alias: 'r',
+										$expr: {
+											r: {
+												is_of__api_key: {
+													$any: {
+														$alias: 'khr',
+														$expr: {
+															khr: {
+																api_key: {
+																	$any: {
+																		$alias: 'k',
+																		$expr: {
+																			k: { key: apiKey }
+																		}
+																	}
+																}
+															}
+														}
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+		return getPermissions(permsFilter)
+	}, {
+		primitive: true,
+		max: env.cache.apiKeys.max,
+		maxAge: env.cache.apiKeys.maxAge,
+	}
+)
+
+export const getApiKeyPermissions = (apiKey: string, callback?: (err: Error | undefined, permissions: string[]) => void): Promise<string[]> =>
+	Promise.try(() => {
+		if (_.isString(apiKey)) {
+			return $getApiKeyPermissions(apiKey)
+		} else {
+			throw new Error('API key has to be a string, got: ' + typeof apiKey)
+		}
+	}).nodeify(callback)
+
+const getApiKeyActorId = memoize(
+	(apiKey: string) =>
+		sbvrUtils.api.Auth.get({
+			resource: 'api_key',
+			passthrough: {
+				req: rootRead
+			},
+			options: {
+				$select: 'is_of__actor',
+				$filter: {
+					key: apiKey
+				}
+			}
+		})
+		.then((apiKeys: AnyObject[]) => {
+			if (apiKeys.length === 0) {
+				throw new PermissionError()
+			}
+			const apiKeyActorID = apiKeys[0].is_of__actor.__id
+			if (apiKeyActorID == null) {
+				throw new Error('API key is not linked to a actor?!')
+			}
+			return apiKeyActorID
+		}),
+	{
+		primitive: true,
+		promise: true,
+		maxAge: env.cache.apiKeys.maxAge,
+	}
+)
+
+const checkApiKey = Promise.method((req: PermissionReq, apiKey: string) => {
+	if (apiKey == null || req.apiKey != null) {
+		return
+	}
+	return getApiKeyPermissions(apiKey)
+	.catch((err) => {
+		console.warn('Error with API key:', err)
+		// Ignore errors getting the api key and just use an empty permissions object
+		return []
+	})
+	.then((permissions) => {
+		req.apiKey = {
+			key: apiKey,
+			permissions: permissions
+		}
+	})
+})
+
+export const customAuthorizationMiddleware = (expectedScheme = 'Bearer') => {
+	if (expectedScheme == null) {
```

Comment:
You have the argument default above -- do you still need to check for null?

Reply:
Yeah, the argument default in ts/js/cs2 does
```ts
expectedScheme = expectedScheme === undefined ? 'Bearer' : expectedScheme
```
but this catches `null` as well to match what defaults in coffeescript 1 do.  I plan to remove this in future but it would be a potentially breaking change and need a major version